libraries/holochain-service-client/src/main/java/org/holochain/androidserviceruntime/holochain_service_client/HolochainServiceClient.kt

```diff
@@ -37,46 +37,49 @@ class HolochainServiceClient(
         }
       }
 
+  /// Connect to the service
+  fun connect() {
+    Log.d(TAG, "connect")
+    val intent = Intent()
+    intent.setComponent(ComponentName(this.servicePackageName, this.serviceClassName))
+    this.activity.bindService(intent, this.mConnection, Context.BIND_ABOVE_CLIENT)
+  }
+
   /// Entire process to setup an app
   ///
-  /// Connect to service, install app, enable app, ensure app websocket
+  /// Check if app is installed.
+  /// If not, install it and (optionally) enable it.
+  /// Then ensure there is an app websocket available for it
   suspend fun setupApp(
       installAppPayload: InstallAppPayloadFfi,
       enableAfterInstall: Boolean
   ): AppAuthFfi {
     Log.d(TAG, "setupApp")
-    this.connect()
-    this.waitForConnectReady()
-
-    if (!this.isAppInstalled(installAppPayload.installedAppId!!)) {
-      this.installApp(installAppPayload)
-
-      if (enableAfterInstall) {
-        this.enableApp(installAppPayload.installedAppId!!)
-      }
+    if (this.mService == null) {
+      throw HolochainServiceNotConnectedException()
     }
 
-    return this.ensureAppWebsocket(installAppPayload.installedAppId!!)
-  }
+    val deferred = CompletableDeferred<AppAuthFfi>()
+    var callbackBinder =
+        object : IHolochainServiceCallbackStub() {
+          override fun setupApp(response: AppAuthFfiParcel) {
+            Log.d(TAG, "setupApp callback")
+            deferred.complete(response.inner)
+          }
+        }
+    this.mService!!.setupApp(callbackBinder, InstallAppPayloadFfiParcel(installAppPayload), enableAfterInstall)
 
-  /// Connect to the service
-  fun connect() {
-    Log.d(TAG, "connect")
-    val intent = Intent()
-    intent.setComponent(ComponentName(this.servicePackageName, this.serviceClassName))
-    this.activity.bindService(intent, this.mConnection, Context.BIND_ABOVE_CLIENT)
+    return deferred.await()
   }
 
-  /// Poll until we are connected to the service, or the timeout has elapsed
-  suspend fun waitForConnectReady(timeoutMs: Long = 100L, intervalMs: Long = 5L) {
-    var elapsedMs = 0L
-    while (elapsedMs <= timeoutMs) {
-      Log.d(TAG, "waitForConnectReady " + elapsedMs)
-      if (this.mService != null) break
-
-      delay(intervalMs)
-      elapsedMs += intervalMs
-    }
+  /// Connect to service, wait for connection to be ready, and setupApp
+  suspend fun connectSetupApp(
```

Comment:
renamed kotlin client & tauri plugin functions `setupApp` -> `connectSetupApp` to clarify that it is both connecting to the service, awaiting the connection, then setting up the app.